Sporadic binding errors for NameEhCacheProvider

Attempting to migrate 2.3.x apps to 2.4.x I keep running into EhCache binding errors like this:

```
1) Error in custom provider, net.sf.ehcache.ObjectExistsException: Cache play already exists
      at play.api.cache.EhCacheModule.play$api$cache$EhCacheModule$$bindCache$1(Cache.scala:181):
    Binding(interface net.sf.ehcache.Ehcache qualified with QualifierInstance(@play.cache.NamedCache(value=play)) to ProviderTarget(play.api.cache.NamedEhCacheProvider@171916b7)) (via modules: com.google.inject.util.Modules$OverrideModule -> play.api.inject.guice.GuiceableModuleConversions$$anon$1)
      while locating net.sf.ehcache.Ehcache annotated with @play.cache.NamedCache(value=play)
      at play.api.cache.EhCacheModule.play$api$cache$EhCacheModule$$bindCache$1(Cache.scala:182):
    Binding(interface play.api.cache.CacheApi qualified with QualifierInstance(@play.cache.NamedCache(value=play)) to ProviderTarget(play.api.cache.NamedCacheApiProvider@50f2c14d)) (via modules: com.google.inject.util.Modules$OverrideModule -> play.api.inject.guice.GuiceableModuleConversions$$anon$1)
      while locating play.api.cache.CacheApi annotated with @play.cache.NamedCache(value=play)
      while locating play.api.cache.CacheApi

1 error
	at com.google.inject.internal.InjectorImpl$2.get(InjectorImpl.java:1025) ~[guice-4.0.jar:na]
	at com.google.inject.internal.InjectorImpl.getInstance(InjectorImpl.java:1051) ~[guice-4.0.jar:na]
	at play.api.inject.guice.GuiceInjector.instanceOf(GuiceInjectorBuilder.scala:321) ~[play_2.11.jar:2.4.1-SNAPSHOT]
	at play.api.inject.guice.GuiceInjector.instanceOf(GuiceInjectorBuilder.scala:316) ~[play_2.11.jar:2.4.1-SNAPSHOT]
	at play.api.Application$$anonfun$instanceCache$1.apply(Application.scala:236) ~[play_2.11.jar:2.4.1-SNAPSHOT]
	at play.api.Application$$anonfun$instanceCache$1.apply(Application.scala:234) ~[play_2.11.jar:2.4.1-SNAPSHOT]
	at play.utils.InlineCache.fresh(InlineCache.scala:69) ~[play_2.11.jar:2.4.1-SNAPSHOT]
	at play.utils.InlineCache.apply(InlineCache.scala:60) ~[play_2.11.jar:2.4.1-SNAPSHOT]
	at play.api.cache.Cache$.cacheApi(Cache.scala:63) ~[play-cache_2.11.jar:2.4.1-SNAPSHOT]
	at play.api.cache.Cache$.getAs(Cache.scala:127) ~[play-cache_2.11.jar:2.4.1-SNAPSHOT]
Caused by: net.sf.ehcache.ObjectExistsException: Cache play already exists
	at net.sf.ehcache.CacheManager.addCache(CacheManager.java:1171) ~[ehcache-core-2.6.11.jar:na]
	at play.api.cache.NamedEhCacheProvider.get$lzycompute(Cache.scala:213) ~[play-cache_2.11.jar:2.4.1-SNAPSHOT]
	at play.api.cache.NamedEhCacheProvider.get(Cache.scala:210) ~[play-cache_2.11.jar:2.4.1-SNAPSHOT]
	at play.api.cache.NamedEhCacheProvider.get(Cache.scala:208) ~[play-cache_2.11.jar:2.4.1-SNAPSHOT]
	at com.google.inject.util.Providers$GuicifiedProvider.get(Providers.java:119) ~[guice-4.0.jar:na]
	at com.google.inject.internal.ProviderInternalFactory.provision(ProviderInternalFactory.java:81) ~[guice-4.0.jar:na]
	at com.google.inject.internal.InternalFactoryToInitializableAdapter.provision(InternalFactoryToInitializableAdapter.java:53) ~[guice-4.0.jar:na]
	at com.google.inject.internal.ProviderInternalFactory.circularGet(ProviderInternalFactory.java:61) ~[guice-4.0.jar:na]
	at com.google.inject.internal.InternalFactoryToInitializableAdapter.get(InternalFactoryToInitializableAdapter.java:45) ~[guice-4.0.jar:na]
	at com.google.inject.internal.InjectorImpl$2$1.call(InjectorImpl.java:1016) ~[guice-4.0.jar:na]
```

This seems to be cropping up mostly when:
 - the app is using the legacy global `Cache` instance
 - running repeated tests, possibly with the same `Application` instance (e.g. with the `WithBrowser` test helper, which takes an app argument.)

It also seems to happen on a somewhat sporadic basis, making it hard to put together a minimal sample app which consistently fails. I'm working on that.

While this error might be pointing to the pitfalls of using global state, I don't think this kind of error should be leaking out. That is, if there's actually a problem here the framework should throw a more useful error.

The immediate cause can be fixed by replacing `manager.addCache(name)` with `manager.addCacheIfAbsent(name)` in the `NamedEhCacheProvider`. I'm happy to put in a PR but because I'm a bit hazy on how and when Guice caches the things it binds, maybe fixing this would mask a more pernicious problem with app state in tests.

--- a/framework/src/play-cache/src/main/scala/play/api/cache/Cache.scala
+++ b/framework/src/play-cache/src/main/scala/play/api/cache/Cache.scala
@@ -145,8 +145,7 @@ trait EhCacheComponents {
    * Use this to create with the given name.
    */
   def cacheApi(name: String): CacheApi = {
-    ehCacheManager.addCache(name)
-    new EhCacheApi(ehCacheManager.getEhcache(name))
+    new EhCacheApi(NamedEhCacheProvider.getNamedCache(name, ehCacheManager))
   }
 
   lazy val defaultCacheApi: CacheApi = cacheApi("play")
@@ -204,9 +203,20 @@ class CacheManagerProvider @Inject() (env: Environment, config: Configuration, l
 
 private[play] class NamedEhCacheProvider(name: String) extends Provider[Ehcache] {
   @Inject private var manager: CacheManager = _
-  lazy val get: Ehcache = {
+  lazy val get: Ehcache = NamedEhCacheProvider.getNamedCache(name, manager)
+}
+
+private[play] object NamedEhCacheProvider {
+  def getNamedCache(name: String, manager: CacheManager) = try {
     manager.addCache(name)
     manager.getEhcache(name)
+  } catch {
+    case e: ObjectExistsException =>
+      throw new EhCacheExistsException(
+        s"""An EhCache instance with name '$name' already exists.
+           |
+           |This usually indicates that multiple instances of a dependent component (e.g. a Play application) have been started at the same time.
+         """.stripMargin, e)
   }
 }
 
@@ -230,6 +240,8 @@ private[play] class NamedCachedProvider(key: BindingKey[CacheApi]) extends Provi
     new Cached(injector.instanceOf(key))
   }
 }
+
+private[play] case class EhCacheExistsException(msg: String, cause: Throwable) extends RuntimeException(msg, cause)
 
 @Singleton
 class EhCacheApi @Inject() (cache: Ehcache) extends CacheApi {

